### PR TITLE
Add subscription-manager support for Hosted or Satellite

### PR DESCRIPTION
**Pull request by vvaldez**
#### What does this PR do?

Adds a subscription-manager role that can support RHSM Hosted or a Satellite server. Authentication can be username/password or an activation key. Repositories can also be passed in as a parameter.
#### How should this be manually tested?

First refer to the subscription-manager role from a playbook, then set the variables appropriate to your scenario. A full README has been included to describe all parameters.

To use a username/password securely with **ansible-vault**:
1. Create a file to contain the variable such as **secrets.yml**:
   
   ```
   ---
   rhsm_password: "my_secret_password"
   # other variables can optionally be placed here as well
   ```
2. Encrypt the file with **ansible-vault**:
   
   ```
   $ ansible-vault encrypt secrets.yml
   Vault password: 
   Confirm Vault password: 
   Encryption successful
   ```
3. When executing **ansible-playbook** specify **--ask-vault-pass** to be prompted for the decryption password, and also specify the location of the **secrets.yml** as such:
   
   ```
   $ ansible-playbook --ask-vault-pass --extra-vars=@secrets.yml --extra-vars="rhsm_username=myusername" <other playbook options>
   ```
#### Is there a relevant Issue open for this?

No known issues.
#### Who would you like to review this?

/cc @oybed @etsauer @sabre1041

### CODE REVIEW

**Comment by oybed** [comment]
Including @etsauer and @sabre1041 for review as well.

File: rhc-ose-ansible/roles/subscription-manager/README.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+To register to RHSM Hosted with username and password:
+
+```
+ansible-playbook -i inventory/ose-provision ose-provision.yml -e "rhsm_method='hosted' rhsm_username=vvaldez rhsm_password='hunter2' openstack_key_name='vvaldez'"
+```
+
+To register to a Satellite server with an activation key:
+
+```
+ansible-playbook -i inventory/ose-provision ose-provision.yml -e "rhsm_server='10.12.32.1' rhsm_org='cloud_practice' rhsm_activationkey='rhel-7-ose-3-1' openstack_key_name='vvaldez' rhsm_method='satellite'"
```
Comment (etsauer) [inline rhc-ose-ansible/roles/subscription-manager/README.md]: @vvaldez probably want to obfuscate environment specific suff here.

**Comment by etsauer** [comment]
@vvaldez can you provide a list of things that need tested? different configuration combinations you see used here?

**Comment by etsauer** [comment]
@vvaldez i've successfully tested the satellite integration. Here's my test:

```
ansible-playbook -i rhc-ose-vvaldez/rhc-ose-ansible/inventory/ose-provision rhc-ose-vvaldez/rhc-ose-ansible/ose-provision.yml -e "rhsm_server='casl-sat6-1.rhc-ose.labs.redhat.com' rhsm_org='RHC-OSE' rhsm_activationkey='OSEv3-Library' openstack_key_name='esauer' rhsm_method='satellite'"
```

the one thing I noticed is that, the inventory file mentions 3 repositories:

`rhsm_repos='["rhel-7-server-rpms", "rhel-7-server-ose-3.1-rpms", "rhel-7-server-extras-rpms"]'`

however, when I go to log into one of the hosts that gets provisioned, I see this:

```
[root@master1-testenv-ydgvcqvm ~]# yum repolist
Loaded plugins: search-disabled-repos
repo id                                                                                  repo name                                                                                         status
!RHC-OSE_Extra_Packages_for_Enterprise_Linux_EPEL_7_-_x86_64                             EPEL 7 - x86_64                                                                                   12,499
!rhel-7-server-ose-3.1-rpms/x86_64                                                       Red Hat OpenShift Enterprise 3.1 (RPMs)                                                              494
!rhel-7-server-rpms/7Server/x86_64                                                       Red Hat Enterprise Linux 7 Server (RPMs)                                                          10,152
!rhel-7-server-satellite-tools-6.1-rpms/x86_64                                           Red Hat Satellite Tools 6.1 (for RHEL 7 Server) (RPMs)                                                79
!rhel-ha-for-rhel-7-server-rpms/7Server/x86_64                                           Red Hat Enterprise Linux High Availability (for RHEL 7 Server) (RPMs)                                190
repolist: 23,414
```

From this, it looks like:
1. Need to --disable="*" to make sure we're not getting anything we didn't ask for
2. We're missing `rhel-7-server-extras-rpms`... why?

**Comment by etsauer** [comment]
So it looks like the repos that I end up with are just the repos that are set as active in my satellite content view. So the enable repos block appears to not be working.

File: rhc-ose-ansible/roles/subscription-manager/README.md (hunk 1 above)
Comment (vvaldez) [inline rhc-ose-ansible/roles/subscription-manager/README.md]: This was just example data, not a real password, but I could obfuscate if needed.

**Comment by vvaldez** [comment]
@etsauer this was intentional, if an activation key was used, the --disable=\* and the --enable of the specified repos tasks are skipped (the README states an activationkey set will ignore any repos specified). This can be changed, but typically repos are set in the key so I didn't want to override any provided repos by the key. Should I remove the checks to skip these tasks (lines 83 and 90 in rhc-ose-ansible/roles/subscription-manager/tasks/main.yml)?

After thinking about this, I could see a scenario where an activation key can be used for authentication instead of username/password, and simply all repos are added to the key. In this scenario I could see disable/enable being used. Perhaps only skip disable/enable if rhsm_repos is empty.

File: rhc-ose-ansible/roles/subscription-manager/README.md (hunk 1 above)
Comment (vvaldez) [inline rhc-ose-ansible/roles/subscription-manager/README.md]: FYI the password is a reference to: http://knowyourmeme.com/memes/hunter2 http://mamamusings.net/archives/2005/04/12/laughing_out_loud.php

**Comment by vvaldez** [comment]
@etsauer what's the best way to provide test scenarios?

So far I see these combinations:

These should succeed:
- Satellite server with activation key
- Satellite server with username and password
- Satellite or Hosted with username and password and specified pool_id
- Satellite with both username/password and activation key specified (key should take precedence)
- Hosted with username and password
- No subscription manager (no variables set)
- Pre-registered system (subscription-manager status should return a 0 which the current image does not do) should skip all registration regardless of what is specified
- Satellite with activation key that specifies repos, with rhsm_repos variable not set (added in latest commit)
- Satellite or Hosted with rhsm_repos set to specific list of repos

These should fail:
- Satellite without an org specified
- Satellite or Hosted with only a username or password set
- Hosted with a Satellite server specified (rhsm_server)
- Hosted with an activation key ~~(this should fail but looking at the code I may not have considered this)~~ added in latest commit
- Test a username or password with special characters (a ! in a password caused me issues if I used '' around the it)
- Invalid username, password, activation key or pool_id has not been tested

**Comment by vvaldez** [comment]
Refactored based on discussion, @etsauer please pull the latest and try your test again. The only thing you need to change is remove "rhsm_method" and rename "rhsm_server" to "rhsm_satellite". 

I changed the conditional I mentioned from activation key to checking if the rhsm_repos variable is set, which will then run disable/enable.This should work as you expected. I also updated my test scenario list.

**Comment by oybed** [comment]
Looks good - I successfully ran the test and get the instances registered to the Satellite server.

Commands executed as part of test:

```
>> git clone <source>
>> cd rhc-ose/rhc-ose-ansible
>> source ~/.openstack/openrc.sh
>> ansible-playbook -i inventory/ose-provision ose-provision.yml -e "rhsm_satellite='casl-sat6-1.rhc-ose.labs.redhat.com' rhsm_org='RHC-OSE' rhsm_activationkey='OSEv3_1' openstack_key_name='obedin-ose3-pub-key'"
```

**Comment by oybed** [comment]
@etsauer @sabre1041 you A_OK if we merge this one?

**Comment by etsauer** [comment]
@oybed someone needs to test the manual subscription piece as well i think.

**Comment by oybed** [comment]
@etsauer good point - doing that right now

File: rhc-ose-ansible/roles/subscription-manager/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  command: "/usr/bin/subscription-manager attach --auto"
+  when:
+    - not registered
+    - rhsm_authentication != "key"
+
+- name: Attach to a specific pool
+  command: "/usr/bin/subscription-manager attach --pool={{ rhsm_pool }}"
+  when:
+    - rhsm_pool is defined and rhsm_pool is not none and rhsm_pool|trim != ''
+    - and not registered
```
Comment (oybed) [inline rhc-ose-ansible/roles/subscription-manager/tasks/main.yml]: this extra "and" is causing problems for runs with username/password for CDN registration.

File: rhc-ose-ansible/roles/subscription-manager/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  when:
+    - not registered
+    - rhsm_authentication == 'key'
+    - rhsm_satellite is defined
+    - rhsm_satellite is not none
+    - rhsm_satellite|trim != ''
+
+# This can apply to either Hosted or Satellite
+- name: Register using username and password
+  command: "/usr/bin/subscription-manager register --username={{ rhsm_username }} --password={{ rhsm_password }}"
```
Comment (oybed) [inline rhc-ose-ansible/roles/subscription-manager/tasks/main.yml:75]: This leaves the username and password in the logs (i.e.: the ansible-command) + potentially in the command history. Hence, anybody using this environment after the fact can obtain this info. For the non-ansible tools, we spent quite a bit of time on ensuring that no username / password is kept/exposed within the image(s), so we'd want to make sure the same is true here.

**Comment by oybed** [comment]
@vvaldez found some additional items to work on during my test of the using username/password case.Please have a look at the comments and let me know if you have any questions. 

Commands executed as part of the test:

```
>> git clone <source>
>> cd rhc-ose/rhc-ose-ansible
>> source ~/.openstack/openrc.sh
>> ansible-playbook -i inventory/ose-provision ose-provision.yml -v -e "rhsm_username='<username>' rhsm_password='<password>' rhsm_pool='<pool-id>' openstack_key_name='obedin-ose3-pub-key'" 
```

File: rhc-ose-ansible/roles/subscription-manager/tasks/main.yml (hunk 2 above)
Comment (vvaldez) [inline rhc-ose-ansible/roles/subscription-manager/tasks/main.yml]: Good catch, I hadn't tested with a pool, will fix.

**Comment by vvaldez** [comment]
I've addressed these two issues in the latest commits. I had a bit of an issue coming up with the best way to securely obfuscate the password and settled on using a private **vars_prompt** and **no_log** setting for the task. This works as intended and masks the password even when using **-vvv**. The password variable is prompted at the start of the playbook and the only way to reveal it in any way is to use a **debug** task inside the role. The only downside is that this variable prompt occurs even if asubscription-manager password will not be used. There is no way around this that I've found as there is no support for conditional **vars_prompt** (see http://stackoverflow.com/questions/25466675/ansible-to-conditionally-prompt-for-a-variable). This seems to be an acceptable compromise and if it becomes an issue for some you can always remove the **vars_prompt** from the ose-provision.yml and just ignore this role. If this is not an acceptable compromise I can research using **ansible-vault** to decrypt an encrypted file to retrieve the password.

Here is a sample run showing the relevant lines:

During a normal playbook run with -v during the registration the CLI output is this:

```
TASK: [subscription-manager | Register using username and password] *********** 
changed: [master1-vvaldez-passwordtest14] => {"censored": "results hidden due to no_log parameter", "changed": true, "rc": 0}
changed: [node1-vvaldez-passwordtest14] => {"censored": "results hidden due to no_log parameter", "changed": true, "rc": 0}
changed: [node2-vvaldez-passwordtest14] => {"censored": "results hidden due to no_log parameter", "changed": true, "rc": 0}
```

Here is a run with -vvv:

```
$ ansible-playbook -vvv -i inventory/ose-provision ose-provision.yml -e "rhsm_username=vvaldez openstack_key_name=vvaldez env_id=vvaldez-passwordtest15 rhsm_pool=8a85f9823cf494b3013dcc6055a041f5"
 [WARNING]: provided hosts list is empty, only localhost is available

Subscription Manager password (use blank if using an Activation Key or to disable registration): 
confirm Subscription Manager password (use blank if using an Activation Key or to disable registration): 
...
TASK: [Set password fact] ***************************************************** 
ok: [localhost] => {"censored": "results hidden due to no_log parameter"}
...
TASK: [subscription-manager | Register using username and password] *********** 
<10.3.14.170> ESTABLISH CONNECTION FOR USER: root
<10.3.14.170> REMOTE_MODULE command (no_log enabled, args censored)
<10.3.14.170> EXEC ssh -C -tt -v -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/home/vvaldez/.ansible/cp/ansible-ssh-%h-%p-%r" -o StrictHostKeyChecking=no -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=10 10.3.14.170 /bin/sh -c 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1459895788.78-42242306788188 && chmod a+rx $HOME/.ansible/tmp/ansible-tmp-1459895788.78-42242306788188 && echo $HOME/.ansible/tmp/ansible-tmp-1459895788.78-42242306788188'
<10.3.14.173> ESTABLISH CONNECTION FOR USER: root
<10.3.14.173> REMOTE_MODULE command (no_log enabled, args censored)
<10.3.14.173> EXEC ssh -C -tt -v -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/home/vvaldez/.ansible/cp/ansible-ssh-%h-%p-%r" -o StrictHostKeyChecking=no -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=10 10.3.14.173 /bin/sh -c 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1459895788.79-94022306550990 && chmod a+rx $HOME/.ansible/tmp/ansible-tmp-1459895788.79-94022306550990 && echo $HOME/.ansible/tmp/ansible-tmp-1459895788.79-94022306550990'
<10.3.14.176> ESTABLISH CONNECTION FOR USER: root
<10.3.14.176> REMOTE_MODULE command (no_log enabled, args censored)
<10.3.14.176> EXEC ssh -C -tt -v -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/home/vvaldez/.ansible/cp/ansible-ssh-%h-%p-%r" -o StrictHostKeyChecking=no -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=10 10.3.14.176 /bin/sh -c 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1459895788.8-29944679084376 && chmod a+rx $HOME/.ansible/tmp/ansible-tmp-1459895788.8-29944679084376 && echo $HOME/.ansible/tmp/ansible-tmp-1459895788.8-29944679084376'
<10.3.14.173> PUT /tmp/tmp8n7BBF TO /root/.ansible/tmp/ansible-tmp-1459895788.79-94022306550990/command
<10.3.14.176> PUT /tmp/tmp2CPTAG TO /root/.ansible/tmp/ansible-tmp-1459895788.8-29944679084376/command
<10.3.14.170> PUT /tmp/tmpZuDqNg TO /root/.ansible/tmp/ansible-tmp-1459895788.78-42242306788188/command
<10.3.14.173> EXEC ssh -C -tt -v -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/home/vvaldez/.ansible/cp/ansible-ssh-%h-%p-%r" -o StrictHostKeyChecking=no -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=10 10.3.14.173 /bin/sh -c 'LANG=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /usr/bin/python /root/.ansible/tmp/ansible-tmp-1459895788.79-94022306550990/command; rm -rf /root/.ansible/tmp/ansible-tmp-1459895788.79-94022306550990/ >/dev/null 2>&1'
<10.3.14.176> EXEC ssh -C -tt -v -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/home/vvaldez/.ansible/cp/ansible-ssh-%h-%p-%r" -o StrictHostKeyChecking=no -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=10 10.3.14.176 /bin/sh -c 'LANG=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /usr/bin/python /root/.ansible/tmp/ansible-tmp-1459895788.8-29944679084376/command; rm -rf /root/.ansible/tmp/ansible-tmp-1459895788.8-29944679084376/ >/dev/null 2>&1'
<10.3.14.170> EXEC ssh -C -tt -v -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/home/vvaldez/.ansible/cp/ansible-ssh-%h-%p-%r" -o StrictHostKeyChecking=no -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=10 10.3.14.170 /bin/sh -c 'LANG=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /usr/bin/python /root/.ansible/tmp/ansible-tmp-1459895788.78-42242306788188/command; rm -rf /root/.ansible/tmp/ansible-tmp-1459895788.78-42242306788188/ >/dev/null 2>&1'
changed: [node1-vvaldez-passwordtest15] => {"censored": "results hidden due to no_log parameter", "changed": true, "rc": 0}
changed: [master1-vvaldez-passwordtest15] => {"censored": "results hidden due to no_log parameter", "changed": true, "rc": 0}
changed: [node2-vvaldez-passwordtest15] => {"censored": "results hidden due to no_log parameter", "changed": true, "rc": 0}
```

**Comment by vvaldez** [comment]
I also see there is a merge conflict, I will address this and update this PR.

**Comment by etsauer** [comment]
@vvaldez i'm thinking if we have to deal with a prompt that can't be conditional, then what we might want to do is split satellite support and subscription-manager support into separate roles. @oybed thoughts?

**Comment by vvaldez** [comment]
Except password authentication can apply to either. Right now if you don't want to set a password you simply hit ENTER twice. We could disable the prompt confirmation which would reduce that to only once.

I can research using an encrypted file using **ansible-vault** which can conditionally load the password only if the file exists and provide instructions outside the playbook to create it. Otherwise if we try to build that workflow into the playbook we're back to the same point of unconditional prompt.

**Comment by oybed** [comment]
I agree with etsauer that it should be possible to execute one of these playbooks without a prompt - e.g: if an activation key is used. For example for cases where the provisioning is fully automated - having an option to avoid the prompt will then make that task easier. 

If going the route of 2 roles, I agree that vvaldez that password authentication can be applied to either, so probably need to look for another logical split - i.e.: password role v.s. other (e.g.: activation key).

**Comment by vvaldez** [comment]
Merge conflict resolved.

I just discovered that if "rhsm_password" is defined at the start of the playbook run (via -e) then it does not prompt for the variable. So one workaround is to set "-e rhsm_password=none" or "=false" and the prompt is skipped. If you also specify "rhsm_activationkey" this already takes precedence over rhsm_password. So this is one way to avoid the prompt without changing this PR, allowing interactive prompting if needed.

Here's an example that does not prompt and successfully registers with an activation key:

```
ansible-playbook -v -i inventory/ose-provision ose-provision.yml -e "rhsm_satellite=satellite.example.com openstack_key_name=vvaldez env_id=vvaldez-test20 rhsm_password=none rhsm_org=cloud_practice rhsm_activationkey=rhel-7-ose-3-1"
```

However, ignoring the above scenario, I propose to simply comment out or remove the vars_prompt from ose-provision.yml, then this role will work out of the box with an activation key for full automation (the primary use case for this environment as I understand) or with no subscription manager used.

The README has instructions on how to add the prompt for interactive password registration should one be needed (or it could be left in and uncommented). If a password is supplied on the CLI it will still work (without a prompt defined or with, as described above it will not prompt if rhsm_password is set) for automation and will be obfuscated in the playbook with the no_log parameter. However, it would still potentially be in the user's command history (the same way it would be in any history executing subscription manager with a password). If this is a concern this is where vars_prompt comes in.

Is this acceptable? I don't see a need to split this into multiple roles, the only difference between password vs. activationkey is adding a vars_prompt if a password is needed with full obfuscation, and that is done at the play level, not the role. 

In summary, since Ansible can't do a conditional prompt, the user has to make that conditional choice and uncomment (or add) a prompt as needed or use the work around of specifying rhsm_password=none on the CLI.

**Comment by vvaldez** [comment]
As per our discussion on the call today I removed the **vars_prompt** and documented how to add it in the README. I tested trying to conditionally include a vars_prompt but include is not allowed before pre_task. I also tested not supplying a password and pasting it in during the task but this echos the password to the user's screen. I settled on using **ansible-vault** to encrypt a file containing the variable and this works. It was much easier than I realized. I've documented how to do this in the README. In fact, any variable can go in this file such as username and activationkey.

The following scenarios work out of the box:
- Username/password on the CLI or in an unencrypted file passed via --extra-vars (not secure but this is difficult to detect and block)
- Username passed via --extra-vars and password in an encrypted file unlocked with --ask-vault-pass:
  
  ```
  ansible-playbook -v -i inventory/ose-provision ose-provision.yml -e @secrets.yml --ask-vault-pass -e "rhsm_username=vvaldez"
  ```
- Activation Key to Satellite

The following scenario fails (intentionally as this was not working as intended and is not scalable, and informs the user about referring to the README to enable one of the two prompt options):
- Username specified but not password

To remedy this scenario requires a copy/paste from the README to enable a prompt:
- Username specified but not password (password gathered via vars_prompt)

Instructions to create an encrypted file containing variables including the password:
1. Create a file to contain the variable such as **secrets.yml**:
   
   ```
   ---
   rhsm_password: "my_secret_password"
   # other variables can optionally be placed here as well
   ```
2. Encrypt the file with **ansible-vault**:
   
   ```
   $ ansible-vault encrypt secrets.yml
   Vault password: 
   Confirm Vault password: 
   Encryption successful
   ```
3. When executing **ansible-playbook** specify **--ask-vault-pass** to be prompted for the decryption password, and also specify the location of the **secrets.yml** as such:
   
   ```
   $ ansible-playbook --ask-vault-pass --extra-vars=@secrets.yml --extra-vars="rhsm_username=myusername" <other playbook options>
   ```

**Comment by oybed** [comment]
LGTM - @etsauer @sabre1041 let me know if you're A_OK with a merge.

**Comment by sabre1041** [comment]
@oybed from a visual inspection it looks good. I am going to try to spin it up in the lab environment to confirm and if all tests pass, we should be good to merge

**Comment by etsauer** [comment]
lgtm

**Comment by vvaldez** [comment]
Added an additional commit to set **ansible_sudo: true** when calling this role. Previous testing was done assuming the root user was being used.

**Comment by etsauer** [comment]
i'm fine with this... can we go ahead and merge? @oybed @sabre1041

**Comment by sabre1041** [comment]
@etsauer good with me
